File: project/backend/app/models/appointment.py
"""
Appointment model for managing appointments and queue system
"""
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import relationship
import enum

from app.db.base import BaseModel


class AppointmentStatus(str, enum.Enum):
    """Appointment status enumeration"""
    SCHEDULED = "scheduled"
    CHECKED_IN = "checked_in"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class AppointmentType(str, enum.Enum):
    """Appointment type enumeration"""
    SCHEDULED = "scheduled"
    WALK_IN = "walk_in"


class Appointment(BaseModel):
    """
    Appointment model for managing appointments and queue system.
    
    Attributes:
        id: Primary key (inherited from Base)
        patient_id: Foreign key to patients table
        doctor_id: Foreign key to doctors table
        scheduled_time: Scheduled appointment time
        status: Current appointment status
        appointment_type: Type of appointment (scheduled or walk-in)
        queue_position: Position in doctor's queue (NULL if not in queue)
        created_at: Timestamp when appointment was created (inherited from Base)
        updated_at: Timestamp when appointment was last updated (inherited from Base)
    
    Relationships:
        patient: Associated patient
        doctor: Associated doctor
        medical_records: Medical records created during this appointment
    """
    __tablename__ = "appointments"
    
    patient_id = Column(Integer, ForeignKey("patients.id", ondelete="CASCADE"), index=True)
    doctor_id = Column(Integer, ForeignKey("doctors.id", ondelete="CASCADE"), index=True)
    scheduled_time = Column(DateTime, nullable=False, index=True)
    status = Column(
        SQLEnum(AppointmentStatus, name="appointment_status", create_type=False),
        default=AppointmentStatus.SCHEDULED,
        nullable=False,
        index=True
    )
    appointment_type = Column(
        SQLEnum(AppointmentType, name="appointment_type_enum", create_type=False),
        default=AppointmentType.SCHEDULED,
        nullable=False
    )
    queue_position = Column(Integer)
    consultation_start_time = Column(DateTime)  # Track when consultation started (IN_PROGRESS)
    
    # Relationships
    patient = relationship("Patient", back_populates="appointments")
    doctor = relationship("Doctor", back_populates="appointments")
    medical_records = relationship("MedicalRecord", back_populates="appointment")
    
    def __repr__(self):
        return f"<Appointment(id={self.id}, patient_id={self.patient_id}, doctor_id={self.doctor_id}, status='{self.status}')>"

File: project/mobile/lib/models/appointment.dart
/// Appointment model representing a scheduled or walk-in appointment
class Appointment {
  final int id;
  final int patientId;
  final int doctorId;
  final DateTime scheduledTime;
  final String status;
  final String appointmentType;
  final int? queuePosition;
  final String? doctorName;
  final String? patientName;
  final DateTime createdAt;

  Appointment({
    required this.id,
    required this.patientId,
    required this.doctorId,
    required this.scheduledTime,
    required this.status,
    required this.appointmentType,
    this.queuePosition,
    this.doctorName,
    this.patientName,
    required this.createdAt,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) {
    return Appointment(
      id: json['id'] as int,
      patientId: json['patient_id'] as int,
      doctorId: json['doctor_id'] as int,
      scheduledTime: DateTime.parse(json['scheduled_time'] as String),
      status: json['status'] as String,
      appointmentType: json['appointment_type'] as String,
      queuePosition: json['queue_position'] as int?,
      doctorName: json['doctor_name'] as String?,
      patientName: json['patient_name'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patient_id': patientId,
      'doctor_id': doctorId,
      'scheduled_time': scheduledTime.toIso8601String(),
      'status': status,
      'appointment_type': appointmentType,
      'queue_position': queuePosition,
      'doctor_name': doctorName,
      'patient_name': patientName,
      'created_at': createdAt.toIso8601String(),
    };
  }

  bool get isScheduled => status == 'scheduled';
  bool get isCheckedIn => status == 'checked_in';
  bool get isInProgress => status == 'in_progress';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
  bool get isWalkIn => appointmentType == 'walk_in';
}
